src/boot/efi/efi-string.c

```diff
@@ -272,6 +274,453 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
 DEFINE_PARSE_NUMBER(char, parse_number8);
 DEFINE_PARSE_NUMBER(char16_t, parse_number16);
 
+enum {
+        SCRATCH_BUF_SIZE = 64,
+};
+
+typedef enum FormatFlags {
+        NUMBER_DEFAULT    = 1 << 0, /* Print as decimal. */
+        NUMBER_HEX        = 1 << 1, /* Print as hex. */
+        NUMBER_HEX_PREFIX = 1 << 2, /* Print as hex with 0x prefix. */
+        NUMBER_LOWER      = 1 << 3, /* Print hex numbers and prefix in lowercase letters. */
+        NUMBER_SIGNED     = 1 << 4, /* Value is signed. */
+} NumberFormatFlags;
+
+static void format_number(
+                char16_t buf[static SCRATCH_BUF_SIZE],
+                size_t min_digits,
+                NumberFormatFlags flags,
+                uint64_t u) {
+
+        size_t n = 0;
+        int base = FLAGS_SET(flags, NUMBER_HEX) ? 16 : 10;
+        char16_t tmp[SCRATCH_BUF_SIZE];
+        const char *digits = FLAGS_SET(flags, NUMBER_LOWER) ? "0123456789abcdef" : "0123456789ABCDEF";
+
+        if (FLAGS_SET(flags, NUMBER_SIGNED) && (int64_t) u < 0) {
```

Comment:
are you sure this cast is safe? isn't this one of these areas where the c compiler can do whatever it wants if u > INT64_MAX because undef behaviour or so?

Reply:
Yeah, it's implementation defined. And for gcc it's two's complement: https://gcc.gnu.org/onlinedocs/gcc/Integers-implementation.html#Integers-implementation
So this should be fine and testcases ought to be covering this anyways.

Also, it's interesting to know gcc only supports two's complement integers.